### PR TITLE
chore: make Count and Export methods idempotent

### PR DESCRIPTION
**Pull request by cmwaters**
This isn't necessary as the expected flow is that these things are only ever called once but nonetheless I think it would be good to ensure that both methods are idempotent i.e. they can be called multiple times with the same result. This is caught me out a couple of times. Even calling css.Count() after css.Export() will give you a different length then if you called it before.

### CODE REVIEW

File: pkg/shares/split_compact_shares.go [hunk 1]
```diff
@@ -213,7 +228,7 @@ func (css *CompactShareSplitter) isEmpty() bool {
 // Count returns the number of shares that would be made if `Export` was invoked
 // on this compact share splitter.
 func (css *CompactShareSplitter) Count() (shareCount int) {
-	if !css.isEmptyPendingShare() {
+	if !css.isEmptyPendingShare() && !css.done {
```
Comment (rootulp) [inline pkg/shares/split_compact_shares.go:231]: I think we can remove the `&& !css.done` here if we wiped the pending share in `.Export()` and added this check

```go
// isEmptyPendingShare returns true if the pending share is empty, false otherwise.
func (css *CompactShareSplitter) isEmptyPendingShare() bool {
	if css.pendingShare == nil {
		return true
	}
...
```

**Review by rootulp — COMMENTED** [review]
Agreed we can make these methods idempotent. I have a few other ideas for improving the clarity compact share splitter:

1. Extract a new pending share helper
2. Modify the `Share` type to use a builder pattern 

but they're unrelated to the changes in this PR

File: pkg/shares/split_compact_shares.go [hunk 2]
```diff
@@ -47,6 +48,14 @@ func (css *CompactShareSplitter) WriteTx(tx coretypes.Tx) {
 
 // WriteBytes adds the delimited data to the underlying compact shares.
 func (css *CompactShareSplitter) WriteBytes(rawData []byte) {
+	if css.done {
+		// remove the last element
+		if !css.isEmptyPendingShare() {
+			css.shares = css.shares[:len(css.shares)-1]
+		}
+		css.done = false
+	}
```
[the remark this answers is not in the document]
Comment (cmwaters) [reply pkg/shares/split_compact_shares.go:57]: I guess it depends what behaviour we want. I was writing this in a way that allowed you to write a tx and then export it and then write another tx and then export and so on. To do so requires us to remove the last share that was added when we ran `Export`.

You're suggestion is also fine. We can modify the behaviour so `WriteTx` can't be called after `Export`. The way I opted for, although being a little complex, means users need to know less about the object; they are free to run `Export`, `Count` and `WriteTx` as many times as they want

[the remark this answers is not in the document]
Comment (rootulp) [reply pkg/shares/split_compact_shares.go:57]: Ahh thanks for explaining the rationale. I'm on-board with modifying the compact share splitter to do this:

> I was writing this in a way that allowed you to write a tx and then export it and then write another tx and then export and so on

Note: If we write a new tx to an already exported share splitter, then the reserved bytes and sequence length also need updating. 
- Based on https://github.com/rootulp/celestia-app/blob/d419b5fa54d6db69903836e292a2eac1f8a21fe1/pkg/shares/split_compact_shares.go#L154 I expect sequence length to be updated. 
- Based on https://github.com/rootulp/celestia-app/blob/d419b5fa54d6db69903836e292a2eac1f8a21fe1/pkg/shares/split_compact_shares.go#L78 I expect reserved bytes to be updated.

The pending share logic isn't intuitive (sorry that's my fault) so do you mind adding a unit test that that verifies the desired functionality? I'm concerned about edge cases with the pending share. For example this test fails:

```go
func TestWriteAfterExport(t *testing.T) {
	partialShare := coretypes.Tx([]byte{1})
	exactlyOneShare := coretypes.Tx(bytes.Repeat([]byte{4}, subtractDelimLen(appconsts.FirstCompactShareContentSize)))
	exactlyTwoShares := coretypes.Tx(bytes.Repeat([]byte{5}, subtractDelimLen(appconsts.FirstCompactShareContentSize+appconsts.ContinuationCompactShareContentSize)))

	css := NewCompactShareSplitter(appconsts.TxNamespaceID, appconsts.ShareVersionZero)

	css.WriteTx(exactlyOneShare)
	shares := css.Export()
	assert.Len(t, shares, 1)

	css.WriteTx(exactlyTwoShares)
	shares = css.Export()
	assert.Len(t, shares, 3)

	css.WriteTx(exactlyOneShare)
	shares = css.Export()
	assert.Len(t, shares, 4)

	css.WriteTx(partialShare)
	shares = css.Export()
	assert.Len(t, shares, 5) // fails
	fmt.Printf("%v", shares)
}

```

[the remark this answers is not in the document]
Comment (cmwaters) [reply pkg/shares/split_compact_shares.go:57]: > Note: If we write a new tx to an already exported share splitter, then the reserved bytes and sequence length also need updating.

Won't they get updated the next time the caller calls `Export`?

[the remark this answers is not in the document]
Comment (cmwaters) [reply pkg/shares/split_compact_shares.go:57]: >  For example this test fails:

I've been trying to debug why it fails. It seems to succeed when the partial share contains 8 or more bytes. So I'm not too sure what it has to do with the logic that I added. To illustrate my point, the following test without the intermediate `Export` calls also fails.

```go
func TestWriteAfterExportV2(t *testing.T) {
	partialShare := coretypes.Tx(bytes.Repeat([]byte{123}, 1))
	exactlyOneShare := generateTx(1)
	exactlyTwoShares := generateTx(2)

	css := NewCompactShareSplitter(appconsts.TxNamespaceID, appconsts.ShareVersionZero)

	css.WriteTx(exactlyOneShare)
	css.WriteTx(exactlyTwoShares)
	css.WriteTx(exactlyOneShare)
	css.WriteTx(partialShare)
	shares := css.Export()
	assert.Len(t, shares, 5) // fails
	fmt.Printf("%v\n", shares)
}
```

The logic to remove the last share (in WriteTx) mimics the condition with which the pending share is added in `Export`. Would appreciate you taking a deeper look

[the remark this answers is not in the document]
Comment (rootulp) [reply pkg/shares/split_compact_shares.go:57]: This test succeeds if I remove the following lines. 

```diff
// WriteBytes adds the delimited data to the underlying compact shares.
func (css *CompactShareSplitter) WriteBytes(rawData []byte) {
-	if css.done {
-		// remove the last element
-		if !css.isEmptyPendingShare() {
-			css.shares = css.shares[:len(css.shares)-1]
-		}
-		css.done = false
-	}
```

I can extract the test case and create a stand-alone PR to `main` so that we have higher confidence that the current implementation behaves as expected. It's def non-intuitive so I'd like to replace the existing pending share logic with a builder pattern for shares.

[the remark this answers is not in the document]
Comment (rootulp) [reply pkg/shares/split_compact_shares.go:57]: ahh I see why now

Screenshot 2023-01-27 at 1 49 46 PM

After these two lines

```go
	css.WriteTx(exactlyOneShare)
	css.WriteTx(exactlyTwoShares)
```

we don't have exactly three shares. Instead, we have 3 shares - 4 bytes because the transaction named `exactlyTwoShares` doesn't need to account for the sequence length written in the first compact share. In other words, our test data was borked. I'll take a stab at writing a clearer test

[the remark this answers is not in the document]
Comment (rootulp) [reply pkg/shares/split_compact_shares.go:57]:
```
func TestWriteAfterExportV2(t *testing.T) {
	type testCase struct {
		name    string
		txs     []coretypes.Tx
		wantLen int
	}
	testCases := []testCase{
		{
			name:    "one tx that occupies exactly one share",
			txs:     []coretypes.Tx{generateTx(1)},
			wantLen: 1,
		},
		{
			name:    "one tx that occupies exactly two shares",
			txs:     []coretypes.Tx{generateTx(2)},
			wantLen: 2,
		},
		{
			name:    "one tx that occupies exactly three shares",
			txs:     []coretypes.Tx{generateTx(3)},
			wantLen: 3,
		},
		{
			name: "two txs that occupy exactly two shares",
			txs: []coretypes.Tx{
				bytes.Repeat([]byte{0xf}, rawTxSize(appconsts.FirstCompactShareContentSize)),
				bytes.Repeat([]byte{0xf}, rawTxSize(appconsts.ContinuationCompactShareContentSize)),
			},
			wantLen: 2,
		},
		{
			name: "three txs that occupy exactly three shares",
			txs: []coretypes.Tx{
				bytes.Repeat([]byte{0xf}, rawTxSize(appconsts.FirstCompactShareContentSize)),
				bytes.Repeat([]byte{0xf}, rawTxSize(appconsts.ContinuationCompactShareContentSize)),
				bytes.Repeat([]byte{0xf}, rawTxSize(appconsts.ContinuationCompactShareContentSize)),
			},
			wantLen: 3,
		},
		{
			name: "four txs that occupy three full shares and one partial share",
			txs: []coretypes.Tx{
				bytes.Repeat([]byte{0xf}, rawTxSize(appconsts.FirstCompactShareContentSize)),
				bytes.Repeat([]byte{0xf}, rawTxSize(appconsts.ContinuationCompactShareContentSize)),
				bytes.Repeat([]byte{0xf}, rawTxSize(appconsts.ContinuationCompactShareContentSize)),
				[]byte{0xf},
			},
			wantLen: 4,
		},
	}
	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			css := NewCompactShareSplitter(appconsts.TxNamespaceID, appconsts.ShareVersionZero)

			for _, tx := range tc.txs {
				css.WriteTx(tx)
			}

			assert.Equal(t, tc.wantLen, css.Count())
			assert.Equal(t, tc.wantLen, len(css.Export()))
		})
	}
}
```

[the remark this answers is not in the document]
Comment (cmwaters) [reply pkg/shares/split_compact_shares.go:57]: Thanks for writing up the test cases. I've added them

**Review by rootulp — APPROVED** [review]

**Review by evan-forbes — APPROVED** [review]

**Review by rootulp — APPROVED** [review]
